### ISSUE TITLE
no support for nested layout templates

### ISSUE THREAD

**Issue by ghost** [marijnh/heckle#7]
I have basic jekyll site that used a layout scheme resembling the following graphs:

    _layouts/default.html : index.html
    _layouts/default.html <- _layouts/post.html : YYYY-MM-DD-whatever.md

There's no support for recursive layout resolution in heckle, requiring some copy-and-paste of some stuff from `default.html` into `post.html`.  I'll have a look at fixing this some other time.

**Comment by colbyrussell** [comment]
````
The following changes since commit 1b21a303bde8e8326db3cd102374d1918004f127:

  Support front matter/layouts for non-markdown docs (2015-05-08 10:19:17 +0200)

are available in the git repository at:

  https://gitlab.com/crussell/heckle.git nestedlayouts

for you to fetch changes up to 8ec6e2c8ce61317d0d49bf08d7f0c9b9783e4e6b:

  Issue #7: Support nested layouts (2016-01-09 15:57:55 -0600)

----------------------------------------------------------------
Colby Russell (1):
      Issue #7: Support nested layouts

 heckle.js | 30 ++++++++++++++++++++----------
 1 file changed, 20 insertions(+), 10 deletions(-)
````

**Comment by marijnh** [comment]
Could you describe how this feature works? It looks like it enables frontmatters for template files, allowing them to have a `layout` field just like Markdown files. What problem does that solve that includes don't solve?

**Comment by colbyrussell** [comment]
Includes aren't as powerful as nested layouts.  (In a similar sense to "regular grammars aren't as powerful as context-free grammars", i.e., where "powerful" is used to talk about something real, and not just an ineffable, qualitative description.)  With includes, you can isolate common pieces, but you still have to go into every layout where you want those elements to appear and add them.  With nested layouts, you just have the subordinate layouts "chain up" to the higher level ones, as below.

Check out the post layout you get when you generate a Jekyll scaffold site.  (The latest version I have is 2.1.0, but I'm not a Rubyist so I'm not going to jerk around with rvm and gem to do whatever it takes to get the latest version working on my machine, and this example should suffice anyway.)

````
colby@ruta-corsica:~$ cd /tmp
colby@ruta-corsica:/tmp$ jekyll new jekyllsite
New jekyll site installed in /tmp/jekyllsite. 
colby@ruta-corsica:/tmp$ cd ./jekyllsite/
colby@ruta-corsica:/tmp/jekyllsite$ cat ./_layouts/post.html 
````
````liquid
---
layout: default
---
<div class="post">

  <header class="post-header">
    <h1>{{ page.title }}</h1>
    <p class="meta">{{ page.date | date: "%b %-d, %Y" }}{% if page.author %} • {{ page.author }}{% endif %}{% if page.meta %} • {{ page.meta }}{% endif %}</p>
  </header>

  <article class="post-content">
  {{ content }}
  </article>

</div>
````
````
colby@ruta-corsica:/tmp/jekyllsite$ cat ./_layouts/default.html 
````
```` liquid
<!DOCTYPE html>
<html>

  {% include head.html %}

    <body>

    {% include header.html %}

    <div class="page-content">
      <div class="wrap">
      {{ content }}
      </div>
    </div>

    {% include footer.html %}

    </body>
</html>
````

Let's say you take that site and want to use it with Heckle (after reworking the templates to use Mold). Some change to Heckle resembling what my patch does is still required in order for it to support this kind of structure, because if not you're copying and pasting the same doctype/html/body/head/header/footer junk from the default layout into post.html.

Note that Jekyll also does templates directly from a document—check out index.html from the same generated skeleton site.  You can't have something like that in Heckle at the moment, either.  The best you can do is create a page-specific layout and then put a dummy postarchives.md in the site root containing only a YAML prologue that references that layout.  Implementing this in Heckle, though, is pretty straightforward if you already have nested layouts working.  See e.g. https://gitlab.com/crussell/schtickle/commit/4c9e52e27d9bc44c3a62099928f4b67d654bcd67 for what it took to handle it in the Heckle-like site generator I'm using for myself.

**Comment by marijnh** [comment]
That makes sense. Merged as 9fdc01c
